src/pydrex/geometry.py

```diff
@@ -0,0 +1,170 @@
+"""> PyDRex: Functions for geometric coordinate conversions and projections."""
+import numpy as np
+from scipy import linalg as la
+
+
+def to_cartesian(ϕ, θ, r=1):
+    """Convert spherical to cartesian coordinates in ℝ³.
+
+    Spherical coordinate convention:
+    - ϕ is the longitude (“azimuth”) ∈ (0, π]
+    - θ is the colatitude (“inclination”) ∈ (0, 2π]
+
+    By default, a radius of `r = 1` is used for the sphere.
+    Returns a tuple containing arrays of x, y, and z values.
+
+    """
+    ϕ = np.atleast_1d(ϕ).astype(float)
+    θ = np.atleast_1d(θ).astype(float)
+    r = np.atleast_1d(r).astype(float)
+    return (r * np.sin(θ) * np.cos(ϕ), r * np.sin(θ) * np.sin(ϕ), r * np.cos(θ))
+
+
+def to_spherical(x, y, z):
+    """Convert cartesian coordinates in ℝ³ to spherical coordinates.
+
+    Spherical coordinate convention:
+    - ϕ is the longitude (“azimuth”) ∈ (0, π]
+    - θ is the colatitude (“inclination”) ∈ (0, 2π]
+
+    Returns a tuple containing arrays of ϕ and θ values.
+
+    """
+    x = np.atleast_1d(x).astype(float)
+    y = np.atleast_1d(y).astype(float)
+    z = np.atleast_1d(z).astype(float)
+    r = np.sqrt(x**2 + y**2 + z**2)
+    return (r, np.arccos(z / r), np.sign(y) * np.arccos(x / np.sqrt(x**2 + y**2)))
+
+
+def poles(orientations, ref_axes="xz", hkl=[1, 0, 0]):
+    """Extract 3D vectors of crystallographic directions from orientation matrices.
+
+    Expects `orientations` to be an array with shape (N, 3, 3).
+    The optional arguments `ref_axes` and `hkl` can be used to change
+    the global reference axes and the crystallographic direction respectively.
+    The reference axes should be given as a string of two letters,
+    e.g. "xz" (default), which specify the second and third axes
+    of the global right-handed reference frame. The third letter in the set "xyz"
+    determines the first axis. The `ref_axes` will therefore become the
+    horizontal and vertical axes of pole figures used to plot the directions.
+
+    """
+    upward_axes = next((set("xyz") - set(ref_axes)).__iter__())
+    axes_map = {"x": 0, "y": 1, "z": 2}
+
+    # Get directions in the right-handed frame.
+    directions = np.tensordot(orientations.transpose([0, 2, 1]), hkl, axes=(2, 0))
+    directions_norm = la.norm(directions, axis=1)
+    directions[:, 0] /= directions_norm
+    directions[:, 1] /= directions_norm
+    directions[:, 2] /= directions_norm
+
+    # Rotate into the chosen reference frame.
+    zvals = directions[:, axes_map[upward_axes]]
+    yvals = directions[:, axes_map[ref_axes[1]]]
+    xvals = directions[:, axes_map[ref_axes[0]]]
+    return xvals, yvals, zvals
+
+
+def lambert_equal_area(xvals, yvals, zvals):
```

Comment:
Is there any similar function in `pyproj`?

Reply:
There is the `laea` projection EPSG code 9820, however the pyproj functions take spherical coord inputs and don't perform the hemisphere flipping. I'm not sure it's worth it to pull that dependency for one function when I would have to do almost the same amount of preprocessing to give it the data in the right conventions. I'm also not sure if they have plain spherical projections rather than using Earth ellipsoids (although it would be surprising if they didn't).